### PR TITLE
build: properly use local npm packages for development and binary build

### PR DESCRIPTION
**Pull request by panzarino**
### MERGE AFTER #14181 
- I thought that some changes there were needed (they ended up not being needed) to make this work so I branched off of that and it's not possible to get this PR in until that (due to squash and merge vs merge commit)

Fixes an issue where public projects would actually resolve to the remote version instead of the local version. Since `*` literally means any version, it would use the version published to npm. In addition, a version of `0.0.0-development` does not technically fall under the `*` rule which translates to `>=0.0.0`.

We can actually use this rule to our advantage. From now on, all local packages will use version `0.0.0-development`, regardless of whether they are public or private (for simplicity, private could stay the same but let's simplify the rules). This ensures that every time we try to use a package that is both local and published to npm, we are always using the local version if it exists. Note: our semantic release tooling will replace `0.0.0-development` with the actual correct version upon deployment to npm.

This also updates the binary build process to use the local version of packages that are also published to npm. This ensures that the binary always ships with exactly what was tested locally and also decouples the binary release from the npm package releases.

Binary build implementation notes: we use yarn to install these packages locally through the `file:...` syntax: for example, server/package.json currently contains `"@cypress/webpack-preprocessor": "0.0.0-development"` which gets replaced with `"@cypress/webpack-preprocessor": "file:../../npm/webpack-preprocessor"`. We do this replacement after copying the npm packages to the dist dir, so that way it only installs the necessary files (the stuff specified in `package.json` that we selectively copy from our private packages already) and doesn't blow up the binary file size. After they get installed to their respective `@packages`, we delete the `npm` directory in the dist dir. Originally I had tried to implement copying + replacing the symlink with a relative path (as we do for `@packages`), but this would not properly work due to peer dependency requirements.

Also resolves a bunch of issues people have been having in development with `@cypress/webpack-preprocessor` which is really the only package right now that is used locally

Note: sometimes when developing locally with public packages that have `peerDependencies` (eg. `webpack-preprocessor`), you'll have to make sure to remove certain `peerDependencies` from  that package's `devDependencies`. For example, `webpack-preprocessor` has a `peerDependency` of `typescript` and for it to work properly it must use the same exact typescript that is used by its dependent (in this case, the binary). By removing it from the `devDependencies` we ensure that it always uses the hoisted version installed at the workspace root, which is the same version that the binary is using. This is essential to ensure that packages are using the same hoisted dependency since installing workspace packages creates a symlink rather than using the same install process as from the npm registry. (This isn't necessary for all `peerDependencies`, just when it involves something like monkey-patching the dependency, etc.)

Right now the PR title has `build:` which will not trigger a new release (even though a couple packages were changed). We can use `fix:` if we want to trigger a release (for npm packages) but since literally nothing public changed this feels redundant

### CODE REVIEW

File: npm/react/package.json [hunk 1]
```diff
@@ -15,7 +15,7 @@
   "dependencies": {
     "@babel/plugin-transform-modules-commonjs": "7.10.4",
     "@cypress/code-coverage": "3.8.1",
```
Comment (lmiller1990) [inline npm/react/package.json:17]: Seems only some packages had their versions updated - should they all be `0.0.0-development`? Why are some `*` and some not?

**Review by lmiller1990 — DISMISSED** [review]
code seems fine, asked 1 q

**Comment by JessicaSachs** [comment]
I'm also a little confused why the mismatch between * and 0.0.0-development -- is there a chance you can update the contributing/monorepo documentation with how to specify local dependencies to explain _why_ we use one vs another?

**Comment by panzarino** [comment]
@lmiller1990 @JessicaSachs So this only applies when we have a name resolution conflict with npm. None of our local private `@packages/` are published to npm, so we can require them with a `*` and everything works fine.

For packages that are published to npm (in our `npm` folder), we need to use this for disambiguation with the npm registry. Since `0.0.0-development` will only exist locally, `yarn` knows to always use this local version.

For package.json:
- private package.json - version doesn't matter (`0.0.0` for convention)
- public package.json - `0.0.0-development`

Requiring other packages:
- private package requiring private package - `"@packages/driver": "*"`
- private package requiring public package - `"@cypress/preprocessor": "0.0.0-development"`
- public package requiring private package - should never do this, won't work
- private package requiring public package - `"@cypress/preprocessor": "0.0.0-development"`

File: npm/react/examples/a11y/yarn.lock [hunk 2]
```diff
@@ -808,7 +808,7 @@
     nyc "15.1.0"
 
 "@cypress/react@file:../../dist":
-  version "0.0.0-development"
+  version "0.0.0-developmentelopment"
```
Comment (lmiller1990) [inline npm/react/examples/a11y/yarn.lock:811]: I am sure you noticed this one already but in case you didnt @panzarino spelling error

Comment (panzarino) [reply npm/react/examples/a11y/yarn.lock:811]: yup - was doing find all & replace and accidentally started modifying the yarn.lock, regenerated lockfile coming when I push

**Comment by panzarino** [comment]
> @lmiller1990 @JessicaSachs So this only applies when we have a name resolution conflict with npm. None of our local private `@packages/` are published to npm, so we can require them with a `*` and everything works fine.
> 
> For packages that are published to npm (in our `npm` folder), we need to use this for disambiguation with the npm registry. Since `0.0.0-development` will only exist locally, `yarn` knows to always use this local version.
> 
> For package.json:
> 
>     * private package.json - version doesn't matter (`0.0.0` for convention)
> 
>     * public package.json - `0.0.0-development`
> 
> 
> Requiring other packages:
> 
>     * private package requiring private package - `"@packages/driver": "*"`
> 
>     * private package requiring public package - `"@cypress/preprocessor": "0.0.0-development"`
> 
>     * public package requiring private package - should never do this, won't work
> 
>     * private package requiring public package - `"@cypress/preprocessor": "0.0.0-development"`

EDIT: We'll be using version `0.0.0-development` for all local packages, regardless of public or private.

File: scripts/npm-release.js [hunk 3]
```diff
@@ -114,7 +114,7 @@ const injectVersions = (packagesToRelease, versions, packages) => {
 
     if (packageJson.dependencies) {
       for (const dependency in packageJson.dependencies) {
-        if (packageJson.dependencies[dependency] === '*') {
+        if (packageJson.dependencies[dependency] === '0.0.0-development' || packageJson.dependencies[dependency] === '*') {
```
Comment (JessicaSachs) [inline scripts/npm-release.js:117]: Why aren't we including the devDeps? Sometimes packages publish with these... right?

File: scripts/npm-release.js [hunk 4]
```diff
@@ -114,7 +114,7 @@ const injectVersions = (packagesToRelease, versions, packages) => {
 
     if (packageJson.dependencies) {
       for (const dependency in packageJson.dependencies) {
```
Comment (JessicaSachs) [inline scripts/npm-release.js:116]: This should be using for...of, right?

**Review by JessicaSachs — DISMISSED** [review]
This looks good! A comment or two here. @lmiller1990 also reviewed ☎️

File: scripts/npm-release.js (hunk 3 above)
Comment (panzarino) [reply scripts/npm-release.js:117]: devDeps doesn't matter for installing from npm, peerDependencies do matter though and we should always put the explicit version number range (even for local packages) directly in the `package.json` (I should probably document this somewhere)

File: scripts/npm-release.js (hunk 4 above)
Comment (panzarino) [reply scripts/npm-release.js:116]: For...in is for properties and for...of is for iterable objects like arrays - [example](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of#Difference_between_for...of_and_for...in)

**Comment by brian-mann** [comment]
Do `npm/*` packages use the local development version of `cypress` when testing / prior to publish?

If so, this is potentially a massive problem. Before publishing `npm/*` packages which rely on `cypress` should only ever publish themselves if they work with the **public** version of cypress. Otherwise we could publish it, and users use it in which the npm package relies on binary API's which are not yet available.

There is no way to automatically solve this - an `npm` package must work on both local binary versions in development **AND** published versions.

**Comment by panzarino** [comment]
@brian-mann Currently yes they do use the local version. However, I don't think the problem that you mention would ever actually happen since npm packages will only be published off of the `master` branch and the `master` branch should always contain the latest release of the binary.

The reason for using the local version is that if you're making changes to an npm package that depend on a new release of the binary (stuff in `develop`), you can do that off of `develop` and the package will be released at the same time as the next binary release. Therefore using the local version and consiously branching off of the correct base (`master` vs `develop`) seems to be a better solution than explicitly binding to `cypress@latest`.

I know we had some discussion around this a while ago and I can look back to see some of our design reasoning at the time.

**Pull request merged**